gems/sorbet-runtime/lib/types/props/decorator.rb

```diff
@@ -11,26 +11,27 @@ class T::Props::Decorator
   extend T::Sig
 
   Rules = T.type_alias {T::Hash[Symbol, T.untyped]}
-  DecoratedClass = T.type_alias {T.untyped} # T.class_of(T::Props), but that produces circular reference errors in some circumstances
-  DecoratedInstance = T.type_alias {T.untyped} # Would be T::Props, but that produces circular reference errors in some circumstances
+  DecoratedInstance = T.type_alias {Object} # Would be T::Props, but that produces circular reference errors in some circumstances
   PropType = T.type_alias {T.any(T::Types::Base, T::Props::CustomType)}
   PropTypeOrClass = T.type_alias {T.any(PropType, Module)}
 
   class NoRulesError < StandardError; end
 
-  sig {params(klass: DecoratedClass).void.checked(:never)}
+  EMPTY_PROPS = T.let({}.freeze, T::Hash[Symbol, Rules])
+  private_constant :EMPTY_PROPS
+
+  sig {params(klass: T.untyped).void}
   def initialize(klass)
-    @class = klass
-    klass.plugins.each do |mod|
-      Private.apply_decorator_methods(mod, self)
+    @class = T.let(klass, T.all(Module, T::Props::ClassMethods))
+    @class.plugins.each do |mod|
+      T::Props::Plugin::Private.apply_decorator_methods(mod, self)
     end
+    @props = T.let(EMPTY_PROPS, T::Hash[Symbol, Rules])
```

Comment:
Couldn't you just do `@props = T.let({}.freeze, T::Hash[Symbol, Rules])` without the need for `EMPTY_PROPS`?

Reply:
I don't think that Ruby is smart enough to avoid constructing a new hash on each initialization that way, even with the `freeze` (AFAIK that only works for strings). So using a constant isn't necessarily an important optimization, but it's easy so why not.

Reply:
Yeah:
```
irb(main):003:0> {}.freeze.object_id
=> 70153547237920
irb(main):004:0> {}.freeze.object_id
=> 70153547256440
irb(main):005:0> ''.freeze.object_id
=> 70153949701740
irb(main):006:0> ''.freeze.object_id
=> 70153949701740
```